### DISCUSSION TITLE
2 vertex spherocylinder definition not supported.

### DISCUSSION THREAD

**Discussion by sauravIISc** [glotzerlab/gsd#310] [Support] [answered]
Dear GSD users,

I am very new to .gsd file format for logging my HOOMD generated trajectories. I am trying to simulate hard particle spherocylinders using SpheroPolyhedron shape definition in HOOMD. I can define spherocylinders only with two vertices for the lower and upper convex hull and the radius using sweep_radius, as also mentioned in HOOMD [documentation](https://hoomd-blue.readthedocs.io/en/latest/module-hpmc-integrate.html#hoomd.hpmc.integrate.ConvexSpheropolyhedron). But when I try to save the shape attributes in a GSD file using a logger.

logger = hoomd.logging.Logger()
logger.add(mc, quantities=['type_shapes'])

It gives me an error about the default number of vertices for the shape SpheroPolyhedron is four not two. Also, OVITO visualisation tool cannot recognise anything with two vertices. Looking forward for the community to help me visualise my results.

Regards,
Saurav.

**Comment by joaander** [comment]
What gives you an error? What is the complete error message? What is the complete (and minimal) script that generates this error message?

All questions regarding OVITO should be directed to OVITO's support channels.

**Reply by sauravIISc** [reply]
Thank you for your prompt reply. Below is the script to generate the initial condition for my system and follow the code for the main simulation.

`#Importing important libraries

import numpy as np
import itertools
import gsd.hoomd

#Simulation parameters

Nrods=900
l=20
radius=0.5
Lx=140
Ly=140
Lz=50

#Initialising my system

spacing=1.5
K=np.ceil(Nrods**(1/2))
L=K*spacing
position= np.zeros((Nrods,3))
x= np.linspace(-L/2,L/2,int(K))
position[:,:2] = np.array(list(itertools.product(x, repeat=2)))
orientation = [(1, 0, 0, 0)] * Nrods

frame = gsd.hoomd.Frame()
frame.particles.N = Nrods
frame.particles.position = position
frame.particles.orientation = orientation
frame.particles.typeid = [0] * Nrods
frame.configuration.box = [Lx, Ly, Lz, 0, 0, 0]
frame.particles.types = ['spherocylinders']
frame.particles.type_shapes=[{"type": "ConvexPolyhedron",'vertices' : [[0.0,0.0,-l/2.0],[0.0,0.0,l/2.0]], "rounding_radius": radius}]
with gsd.hoomd.open(name='initial.gsd', mode='x') as f:
    f.append(frame)
`

After we have initial.gsd we can use it in our actual simulation as below:

`#Importing important libraries

import hoomd

l=20
radius=0.5

#Running the simulation

gpu = hoomd.device.GPU()
simulation = hoomd.Simulation(device=gpu, seed=20)
mc= hoomd.hpmc.integrate.ConvexSpheropolyhedron() 
simulation.create_state_from_gsd(filename='initial.gsd')
mc.shape['spherocylinders']=dict(vertices=[[0.0,0.0,-l/2.0],[0.0,0.0,l/2.0]], sweep_radius=radius)

simulation.operations.integrator = mc

logger = hoomd.logging.Logger()
logger.add(mc, quantities=['type_shapes'])

gsd_writer = hoomd.write.GSD(filename='run.gsd',
                             trigger=hoomd.trigger.Periodic(2),
                             mode='xb',filter=hoomd.filter.All(),
                             logger=logger)
simulation.operations.writers.append(gsd_writer)

simulation.run(20)

gsd_writer.flush()
`
![Shape_error](https://github.com/glotzerlab/gsd/assets/110804394/180e564c-339a-4779-8745-80734a8a4aac)


I have attached the screenshot of the error message I get. I apologize for not sharing the scripts and message beforehand. Thanking you.

Regards,
Saurav.

**Reply by joaander** [reply]
Thanks for the script. For future reference, it is *extremely* helpful to copy and paste the error as text as well. Screenshots of text are very hard to read and impossible to copy and paste from.

**Comment by joaander** [comment, marked as answer]
The error is generated here: https://github.com/glotzerlab/hoomd-blue/blob/4bc6e20d15b83d2a1bf509351c217dc375f7a5e8/hoomd/hpmc/ShapeSpheropolyhedron.h#L195

You are welcome to submit a pull request that removes the error so that two vertex polyhedra can be converted to `type_shapes`. In the meantime, you can work around the issue by including a third vertex on the line connecting the other two (e.g. at the origin).

**Comment by stukowski** [comment]
Just to let you know: We are in the process of addressing the limitation on the OVITO side, see https://gitlab.com/stuko/ovito/-/issues/233

**Reply by joaander** [reply]
Thanks! #1898 removed the restriction in HOOMD-blue - a fix released in 4.9.0.
